haskell/src/Lorentz/Contracts/Stablecoin.hs

```diff
@@ -419,3 +410,101 @@ registryContract =
               $(embedStringFile metadataRegistryContractPath)
         |]
   )
+
+----------------------------------------------------------------------------
+-- TZIP-16 Metadata
+----------------------------------------------------------------------------
+
+metadataMap :: MetadataMap BigMap
+metadataMap = BigMap $ Map.fromList
+  -- One might reasonable expect that the URI would be stored as packed Michelson strings,
+  -- but the TZIP-16 spec is explicit about that not being the case.
+  --
+  -- > Unless otherwise-specified, the encoding of the values must be the direct stream
+  -- > of bytes of the data being stored. (...)
+  -- > There is no implicit conversion to Michelson's binary format (PACK) nor
+  -- > quoting mechanism.
+  --
+  -- See: <https://gitlab.com/tzip/tzip/-/blob/eb1da57684599a266334a73babd7ba82dbbbce66/proposals/tzip-16/tzip-16.md#contract-storage>
+  --
+  -- So, instead, we encode it as UTF-8 byte sequences.
+  [ (mempty, encodeUtf8 @Text "tezos-storage:metadataJSON")
+  , ([mt|metadataJSON|], BSL.toStrict (J.encode metadataJSON))
+  ]
+
+metadataJSON :: Metadata (ToT Storage)
+metadataJSON =
+  Metadata
+    { mName = Just "stablecoin"
+    , mDescription = Nothing
+    , mVersion = Just (toText $ showVersion version)
+    , mLicense = Just License { lName = "MIT", lDetails = Nothing }
+    , mAuthors =
+        [ "Serokell <https://serokell.io/>"
+        , "TQ Tezos <https://tqtezos.com/>"
+        ]
+    , mHomepage = Just "https://github.com/tqtezos/stablecoin/"
+    , mSource = Just Source
+        { sLocation = "https://github.com/tqtezos/stablecoin/tree/v" <> toText (showVersion version) <> "/ligo/stablecoin"
+        , sTools = [ "ligo " ] -- TODO: add ligo version
+        }
+    , mInterfaces = [ "TZIP-12", "TZIP-17" ]
+    , mErrors =
+        [ mkError [mt|XTZ_RECEIVED|]               [mt|Contract received a non-zero amount of tokens and should not proceed any further|]
+        , mkError [mt|NOT_CONTRACT_OWNER|]         [mt|Authorized sender is not contract owner|]
+        , mkError [mt|NOT_PENDING_OWNER|]          [mt|Authorized sender is not current contract pending owner|]
+        , mkError [mt|NO_PENDING_OWNER_SET|]       [mt|Thrown when trying to authorize as pending owner whilst is not set for a contract|]
```

Comment:
> whilst is not set for a contract

Is something missing here?

Reply:
Yea, seems to be :/ I took these messages from the spec, but they need to be revised.

In fact, I think the purpose of having these in the TZIP-16 metadata is so that tools can translate error codes like `NO_PENDING_OWNER_SET`, and show a human-readable message to the user. So these messages should be written with the end-user in mind.

I'll revise them now and submit a new commit.

Reply:
Done, please review the new messages

Reply:
Looks good.